File: src/Robustsort/Bytesort.ts
type Bit = number

type Byte = Bit[]

interface Bytemap {
    pointers: number[],
    data: Byte[],
}

type Metabyte = Bytemap | Metabytemap

interface Metabytemap {
    pointers: number[],
    data: Metabyte[],
}

interface Package {
    reference: number,
    bit: Bit
}

type Bytestack = Metabyte

type Stackmap = Metabytemap

// Dummy function in place of a true bubblesort
const bubblesort = (a: number[]) => a.sort()

// Dummy function in place of a true bubblesort
const bubblesortPackages = (packages: Package[]) => packages.sort((a, b) => a.bit - b.bit)

const bytesort4Bit = (a: number[]) => bytesort(a, 4)

const randomizeArray = (a: number[]): number[] => {
    const newArray: number[] = []
    for (const x of a) {
        if (Math.random() > 0.5) {
            newArray.push(x)
        } else {
            newArray.unshift(x)
        }
    }
    return newArray
}

const bytesort = (bits: number[], bytesize: number): number[] => {
    bits = randomizeArray(bits)
    const bytes = convertRawBitsToBytes(bits, bytesize)
    let bytestacks = getBytestacksFromBytes(bytes)
    while (bytestacks.length > bytesize) {
        bytestacks = reduceBytestacks(bytestacks, 4)
    }
    return getSortedArrayFromBytestacks(bytestacks)
}

const convertRawBitsToBytes = (bits: number[], bytesize: number): Byte[] => {
    const bytes: Byte[] = []
    while (bits.length > 0) {
        let byte = bits.slice(0, bytesize)
        byte = bubblesort(byte)
        bytes.push(byte)
        bits = bits.slice(bytesize)
    }
    return bytes
}

const getBytestacksFromBytes = (bytes: Byte[]): Bytestack[] => {
    const bytestacks: Bytestack[] = []
    while (bytes.length > 0) {
        bytestacks.push(getBytemapFromBytes(bytes.slice(0, 4)))
        bytes = bytes.slice(4)
    }
    return bytestacks
}

const getBytemapFromBytes = (bytes: Byte[]): Bytemap => {
    const pointers: number[] = []
    let i = 0
    while (i < bytes.length) {
        pointers.push(i)
        i++
    }
    return { pointers, data: bytes }
}

const reduceBytestacks = (bytestacks: Bytestack[], bytesize: number): Bytestack[] => {
    const newBytestacks: Bytestack[] = []
    while (bytestacks.length > 0) {
        const bytestack = createBytestack(bytestacks.slice(0, bytesize))
        newBytestacks.push(bytestack)
        bytestacks = bytestacks.slice(bytesize)
    }
    return newBytestacks
}

const createBytestack = (metabytes: Metabyte[]): Bytestack => {
    const packages = getPackagesFromMetabytes(metabytes)
    let packagesSorted = bubblesortPackages(packages)
    const pointers: number[] = []
    for (const package of packagesSorted) {
        pointers.push(package.reference)
    }
    return { pointers, data: metabytes }
}

const getPackagesFromMetabytes = (metabytes: Metabyte[]): Package[] => {
    const packages: Package[] = []
    let i = 0
    for (const metabyte of metabytes) {
        packages.push({ reference: i, bit: getTopValueFromBytestack(metabyte)! })
    }
    return packages
}

const getTopValueFromBytestack = (bytestack: Bytestack): number | null => {
    const topByteOrMetabyte: Byte | Metabyte = bytestack.data[bytestack.pointers[bytestack.pointers.length - 1]]
    switch (topByteOrMetabyte[0]) {
        case "undefined": {
            return null
        }
        case "number": {
            const topByte = topByteOrMetabyte as Byte
            return topByte[topByte.length - 1]
        }
        default: return getTopValueFromBytestack(topByteOrMetabyte as Metabyte)
    }
}

const getSortedArrayFromBytestacks = (bytestacks: Bytestack[]): number[] => {
    const sortedArray: number[] = []
    while (!(areAllEmpty(bytestacks))) {
        const { nextElement, newBytestacks } = getNextElementFromBytestacks(bytestacks)
        sortedArray.unshift(nextElement)
        bytestacks = newBytestacks
    }
    return sortedArray
}

const areAllEmpty = (metabytes: Metabyte[]): boolean => {
    let allEmpty = true
    for (const metabyte of metabytes) {
        if (getTopValueFromBytestack(metabyte) !== null) {
            allEmpty = false
            break
        }
    }
    return allEmpty
}

const getNextElementFromBytestacks = (bytestacks: Bytestack[]): { nextElement: number, newBytestacks: Bytestack[] } => {
    const packages = getPackagesFromMetabytes(bytestacks)
    const sortedPackages = bubblesortPackages(packages)
    const topBytestackIndex = getTopPackage(sortedPackages).reference
    const topBytestack: Bytestack = bytestacks[topBytestackIndex]
    const result = removeTopValueFromBytestack(topBytestack)
    const topValue: number = result.topValue
    const newBytestack: Bytestack[] = [result.newBytestack]
    const newBytestacks: Bytestack[] = bytestacks.slice(0, topBytestackIndex).concat(newBytestack, bytestacks.slice(topBytestackIndex + 1))
    return { nextElement: topValue, newBytestacks }
}

const getTopPackage = (packages: Package[]): Package => {
    while (packages.length > 0) {
        let package = packages.pop()!
        if (typeof package.bit == "number") {
            return package
        }
    }
    throw Error("No package in input contains a numerical value")
}

const removeTopValueFromBytestack = (bytestack: Bytestack): { topValue: number, newBytestack: Bytestack } => {
    const topPointer = bytestack.pointers[bytestack.pointers.length - 1]
    const topByteOrMetabyte: Byte | Metabyte = bytestack.data[topPointer]
    switch (topByteOrMetabyte[0]) {
        // This case should be unnecessary now
        // case "undefined": {
        //     bytestack.pointers.pop()
        //     return removeTopValueFromBytestack(bytestack)
        // }
        case "number": {
            let topByte = topByteOrMetabyte as Byte
            const topValue = topByte.pop()!
            if (topByte.length > 0) {
                topByte = bubblesort(topByte)
            } else {
                bytestack.pointers.pop()
            }
            return { topValue, newBytestack: bytestack }
        }
        default: return removeTopValueFromBytestack(topByteOrMetabyte as Bytestack)
    }
}

File: src/Data/Robustsort/Bytesort.hs
module Data.Robustsort.Bytesort (bytesort, bytesort4Bit) where

import Data.Robustsort.Subalgorithms.Bubblesort (bubblesort)
import Data.Robustsort.Utils.Bytes (BytesortProps, convertRawBitsToBytes, getBytestacksFromBytes, getSortedBitsFromMetastack, mkBSProps, reduceBytestacks)
import Data.Robustsort.Utils.RandomizeList (randomizeList)

bytesort4Bit :: [Int] -> [Int]
bytesort4Bit xs = bytesort xs (mkBSProps 4 bubblesort)

-- | Sort a list of Ints using the Bytesort algorithm

-- | ==== __Examples__
-- >>> bytesort [14, 7, 38, 29, 56, 23, 42, 69, 50, 20, 11, 44, 17, 62, 35, 2, 47, 26, 59, 32] 2
-- [2,7,11,14,17,20,23,26,29,32,35,38,42,44,47,50,56,59,62,69]
-- >>> bytesort [14, 7, 38, 29, 56, 23, 42, 69, 50, 20, 11, 44, 17, 62, 35, 2, 47, 26, 59, 32] 4
-- [2,7,11,14,17,20,23,26,29,32,35,38,42,44,47,50,56,59,62,69]
-- >>> bytesort [14, 7, 38, 29, 56, 23, 42, 69, 50, 20, 11, 44, 17, 62, 35, 2, 47, 26, 59, 32] 8
-- [2,7,11,14,17,20,23,26,29,32,35,38,42,44,47,50,56,59,62,69]
-- >>> bytesort (randomizeList [1..100] 143) 2
-- [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100]
bytesort :: [Int] -> BytesortProps -> [Int]
bytesort xs bsProps = do
  let bits = randomizeList xs 143
  let bytes = convertRawBitsToBytes bits bsProps
  let bytestacks = getBytestacksFromBytes bytes bsProps
  let metastack = reduceBytestacks bytestacks bsProps
  getSortedBitsFromMetastack metastack
